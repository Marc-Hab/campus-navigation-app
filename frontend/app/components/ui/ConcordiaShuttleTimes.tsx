import { ScrollView, View, Text, StyleSheet } from "react-native";
import React, { useState, useEffect } from "react";

// Separate schedules for Loyola and SGW
const loyolaSchedule = [
  "9:15", "9:30", "9:45", "10:00", "10:15", "10:30", "10:45", "11:00",
  "11:15", "11:30", "11:45", "12:30", "12:45", "13:00", "13:15", "13:30",
  "13:45", "14:00", "14:15", "14:30", "14:45", "15:00", "15:15", "15:30",
  "15:45", "16:30", "16:45", "17:00", "17:15", "17:30", "17:45", "18:00",
  "18:15"
];

const sgwSchedule = [
  "9:30", "9:45", "10:00", "10:15", "10:30", "10:45", "11:00", "11:15",
  "11:30", "12:15", "12:30", "12:45", "13:00", "13:15", "13:30", "13:45",
  "14:00", "14:15", "14:30", "14:45", "15:00", "15:15", "15:30", "16:00",
  "16:15", "16:45", "17:00", "17:15", "17:30", "17:45", "18:00", "18:15",
  "18:30"
];

const timeToMinutes = (time: string): number => {
  const [hours, minutes] = time.split(":").map(Number);
  return hours * 60 + minutes;
};


const getCurrentTimeInMinutes = (): number => {
  return timeToMinutes("10:15"); // Test mode: Fixed at 10:15 AM
  // return timeToMinutes(
  //   new Date().toLocaleTimeString("en-US", { hour12: false })
  // );
};

const getNextDepartures = (currentMinutes: number) => {
  const upcomingLoyola = loyolaSchedule
    .filter(time => {
      const timeInMinutes = timeToMinutes(time);
      return timeInMinutes > currentMinutes - 1;
    })
    .slice(0, 3);

  const upcomingSGW = sgwSchedule
    .filter(time => {
      const timeInMinutes = timeToMinutes(time);
      return timeInMinutes > currentMinutes - 1;
    })
    .slice(0, 3);

  return { loyola: upcomingLoyola, sgw: upcomingSGW };
};

export default function ConcordiaShuttleTimes() {
  const [simulatedTime, setSimulatedTime] = useState(getCurrentTimeInMinutes());
  const [departures, setDepartures] = useState(getNextDepartures(simulatedTime));

  useEffect(() => {
    const interval = setInterval(() => {
      setSimulatedTime((prev) => prev + 1);
      setDepartures(getNextDepartures(simulatedTime));
    }, 30000);

    return () => clearInterval(interval);
  }, [simulatedTime]);

  const formatTime = (minutes: number): string => {
    const hours = Math.floor(minutes / 60);
    const mins = Math.floor(minutes % 60);
    return `${hours}:${mins.toString().padStart(2, "0")}`;
  };

  return (
    <ScrollView 
      style={styles.timecontainer}
      contentContainerStyle={styles.contentContainer}
      showsVerticalScrollIndicator={false}
    >
      <Text style={styles.title}>Bus Shuttle Departure Times</Text>
      <Text style={styles.simulatedTime}>
        Current Time: {formatTime(simulatedTime)}
      </Text>
      
      <View style={styles.headerRow}>
        <Text style={styles.headerText}>LOYOLA CAMPUS</Text>
        <Text style={styles.headerText}>SGW CAMPUS</Text>
      </View>

      {[...Array(Math.max(
        departures.loyola.length,
        departures.sgw.length
      ))].map((_, index) => {
        const loyTime = departures.loyola[index];
        const sgwTime = departures.sgw[index];
        const loyMinutesAway = loyTime ? timeToMinutes(loyTime) - simulatedTime : null;
        const sgwMinutesAway = sgwTime ? timeToMinutes(sgwTime) - simulatedTime : null;

        return (
          <View key={index} style={styles.row}>
            <View style={styles.column}>
              {loyTime && (
                <Text style={styles.timeText}>
                  {loyTime}
                  <Text style={[
                    styles.statusText,
                    loyMinutesAway === 0 ? styles.departingStatus : 
                    loyMinutesAway < 0 ? styles.departedStatus : styles.awayStatus
                  ]}>
                    {loyMinutesAway === 0 
                      ? "  ‚è≥ Departing..."
                      : loyMinutesAway < 0 
                        ? "  ‚úì Departed"
                        : `  üöå ${Math.max(0, Math.ceil(loyMinutesAway))} mins away`}
                  </Text>
                </Text>
              )}
            </View>
            <View style={styles.column}>
              {sgwTime && (
                <Text style={styles.timeText}>
                  {sgwTime}
                  <Text style={[
                    styles.statusText,
                    sgwMinutesAway === 0 ? styles.departingStatus : 
                    sgwMinutesAway < 0 ? styles.departedStatus : styles.awayStatus
                  ]}>
                    {sgwMinutesAway === 0
                      ? "  ‚è≥ Departing..."
                      : sgwMinutesAway < 0
                        ? "  ‚úì Departed"
                        : `  üöå ${Math.max(0, Math.ceil(sgwMinutesAway))} mins away`}
                  </Text>
                </Text>
              )}
            </View>
          </View>
        );
      })}
    </ScrollView>
  );
}

const styles = StyleSheet.create({
  timecontainer: {
    backgroundColor: "#1a365d",
    width: "100%",
    maxHeight: "30%", // Keep the container height
  },
  contentContainer: {
    padding: 15,
    paddingBottom: 20, // Extra padding for scroll space
  },
  title: {
    color: "white",
    fontSize: 18,
    fontWeight: "bold",
    textAlign: "center",
    marginBottom: 8,
    textTransform: "uppercase",
  },
  simulatedTime: {
    color: "#9fafca", // Lighter blue for secondary text
    fontSize: 16,
    textAlign: "center",
    marginBottom: 15,
  },
  headerRow: {
    flexDirection: "row",
    justifyContent: "space-around",
    marginBottom: 15,
    borderBottomWidth: 1,
    borderBottomColor: "#2d4a7c",
    paddingBottom: 10,
    paddingHorizontal: 20,
  },
  headerText: {
    color: "white",
    fontSize: 14,
    fontWeight: "bold",
    flex: 1,
    textAlign: 'center',
  },
  row: {
    flexDirection: "row",
    justifyContent: "space-around",
    marginVertical: 8,
    paddingHorizontal: 20,
  },
  column: {
    flex: 1,
    alignItems: 'center',
  },
  timeText: {
    color: "white",
    fontSize: 14,
    fontWeight: "500",
    textAlign: 'center',
    minWidth: 150,
    flexDirection: 'row',
    justifyContent: 'center',
  },
  statusText: {
    fontSize: 14,
    marginLeft: 8,
    textAlign: 'left',
    flex: 1,
  },
  departingStatus: {
    color: "#ffd700", // Gold color for departing
    fontWeight: "bold",
  },
  departedStatus: {
    color: "#48bb78", // Green color for departed
    fontWeight: "bold",
  },
  awayStatus: {
    color: "#9fafca", // Light blue for minutes away
  },
});
