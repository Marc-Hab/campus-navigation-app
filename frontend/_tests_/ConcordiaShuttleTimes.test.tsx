import React from "react";
import { render, screen, fireEvent, act } from "@testing-library/react-native";
import ConcordiaShuttleTimes from "@/app/components/ui/ConcordiaShuttleTimes";
import {
  timeToMinutes,
  getCurrentTimeInMinutes,
  getNextDepartures,
  formatTime,
} from "@/app/components/ui/ConcordiaShuttleTimes";

//timeToMinutes function
describe("timeToMinutes", () => {
  test("converts a string of the form 10:00 to the number of minutes since midnights", () => {
    expect(timeToMinutes("10:00")).toBe(600);
  });

  test("converts a string of the form 0:00 to the number of minutes since midnights", () => {
    expect(timeToMinutes("00:00")).toBe(0);
  });
  test("converts 12:00 to 720 minutes", () => {
    expect(timeToMinutes("12:00")).toBe(720);
  });
  test("returns NaN for invalid time format", () => {
    expect(timeToMinutes("invalid")).toBeNaN();
  });
  test("returns NaN for out-of-range hours", () => {
    expect(timeToMinutes("25:30")).toBeNaN();
  });
});

//getCurrentTimeInMinutes function
describe("getCurrentTimeInMinutes", () => {
  //first of all mock the date object
  beforeAll(() => {
    //The fixed time is set to 17:00
    jest.spyOn(global, "Date").mockImplementation(() => {
      return {
        toLocaleTimeString: () => "17:00", // Mocked time
      };
    });
  });

  afterAll(() => {
    // Restore the original implementation of Date after tests are done
    jest.restoreAllMocks();
  });

  test("returns the mocked current time in minutes", () => {
    // will pass 17:00 to timeToMinutes function
    expect(getCurrentTimeInMinutes()).toBe(1020);
  });
});

//test getDepartures function
describe("getNextDepartures", () => {
  test("returns the next 3 departures after 17:01", () => {
    const currentMinutes = 1021; // 17:01
    expect(getNextDepartures(currentMinutes)).toEqual({
      loyola: ["17:15", "17:30", "17:45"],
      sgw: ["17:15", "17:30", "17:45"],
    });
  });
  test("returns the next 3 departures after 00:00", () => {
    const currentMinutes = 0; // 00:00
    expect(getNextDepartures(currentMinutes)).toEqual({
      loyola: ["9:15", "9:30", "9:45"],
      sgw: ["9:30", "9:45", "10:00"],
    });
  });
});
//formatTime
describe("formattime", () => {
  test("Formats the time from minutes to  a string of the form 0:00", () => {
    expect(formatTime(600)).toBe("10:00");
  });
  test("Formats the time from minutes to  a string of the form 0;00", () => {
    expect(formatTime(0)).toBe("0:00");
  });
  test("Formats the time from minutes to  a string of the form 0:00", () => {
    expect(formatTime(720)).toBe("12:00");
  });
  test("Formats the time from minutes to  a string of the form 0;00", () => {
    expect(formatTime(1441)).toBe("0:01");
  });
});

//test that the components renders correctly
describe("ConcordiaShuttleTimes rendering", () => {
  test("renders the shuttle departure times screen correctly", () => {
    // Renter the parent component
    render(<ConcordiaShuttleTimes />);

    //Make sure the title "Bus Shuttle Departure Times" is rendered
    expect(screen.getByText("Bus Shuttle Departure Times")).toBeTruthy();

    // Make sure the current time is rendering
    expect(screen.getByText(/Current Time:/)).toBeTruthy();

    // Are the row header rendering correctly
    expect(screen.getByText("LOYOLA CAMPUS")).toBeTruthy();
    expect(screen.getByText("SGW CAMPUS")).toBeTruthy();
  });
});
