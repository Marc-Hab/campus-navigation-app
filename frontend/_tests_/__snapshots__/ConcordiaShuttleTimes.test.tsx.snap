// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConcordiaShuttleTimes matches snapshot 1`] = `
<RCTScrollView
  contentContainerStyle={
    {
      "padding": 15,
      "paddingBottom": 20,
    }
  }
  showsVerticalScrollIndicator={false}
  style={
    {
      "backgroundColor": "rgba(145, 35, 55, 0.99)",
      "maxHeight": "30%",
      "width": "100%",
    }
  }
>
  <View>
    <Text
      style={
        {
          "color": "white",
          "fontSize": 18,
          "fontWeight": "bold",
          "marginBottom": 8,
          "textAlign": "center",
          "textTransform": "uppercase",
        }
      }
    >
      Bus Shuttle Departure Times
    </Text>
    <Text
      style={
        {
          "color": "white",
          "fontSize": 16,
          "marginBottom": 15,
          "textAlign": "center",
        }
      }
    >
      Current Time: 
      18:30
    </Text>
    <View
      style={
        {
          "borderBottomColor": "white",
          "borderBottomWidth": 1,
          "flexDirection": "row",
          "justifyContent": "space-around",
          "marginBottom": 15,
          "paddingBottom": 10,
          "paddingHorizontal": 20,
        }
      }
    >
      <Text
        style={
          {
            "color": "white",
            "flex": 1,
            "fontSize": 14,
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      >
        LOYOLA CAMPUS
      </Text>
      <Text
        style={
          {
            "color": "white",
            "flex": 1,
            "fontSize": 14,
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      >
        SGW CAMPUS
      </Text>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
          "justifyContent": "space-around",
          "marginVertical": 8,
          "paddingHorizontal": 20,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flex": 1,
          }
        }
      />
      <View
        style={
          {
            "alignItems": "center",
            "flex": 1,
          }
        }
      >
        <Text
          style={
            {
              "color": "white",
              "flexDirection": "row",
              "fontSize": 14,
              "fontWeight": "500",
              "justifyContent": "center",
              "minWidth": 150,
              "textAlign": "center",
            }
          }
        >
          18:30
          <Text
            style={
              [
                {
                  "flex": 1,
                  "fontSize": 14,
                  "marginLeft": 8,
                  "textAlign": "left",
                },
                {
                  "color": "#ffd700",
                  "fontWeight": "bold",
                },
              ]
            }
          >
              ‚è≥ Departing...
          </Text>
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
